### PR TITLE
Add extension.name clarification

### PR DESCRIPTION
**Pull request by Crissman**
Addressing issue #2100

### CODE REVIEW

**Comment by kmaehashi** [comment]
@beam2d What was the motivation to introduce the concept of `default_name`? (in https://github.com/chainer/chainer/pull/1285/commits/09cc404d1c80a0daac778fbe5a7b3ecea93eb380)

**Comment by niboshi** [comment]
Both `name` and `default_name` attributes are undocumented. How about documenting them and linking to them?

**Comment by beam2d** [comment]
@kmaehashi It is the default value provided as a *property*. I wanted to override the behavior with an attribute, so I used `default_name` instead of `name` (note that we cannot override a property with an attribute with the same name due to the resolution order of `getattr`). It could be resolved in a different manner without using the name `default_name` (e.g. also writing a property setter to emulate attribute-like behavior, using non-data descriptor), but I did not choose them because I felt them a bit trickier.

**Comment by Crissman** [comment]
@niboshi `default_name` is mentioned here. Is it enough to link to this page? https://docs.chainer.org/en/stable/reference/generated/chainer.training.Extension.html#chainer.training.Extension

**Comment by kmaehashi** [comment]
@beam2d Thanks for clarification. I understood that this feature is basically for users who want to override the default name quickly (e.g., by calling `ext.name = 'MyExtension'` before calling `Trainer.extend`).

@Crissman I think @niboshi -san's comment includes the following:

* Add `name = None` to `Extension` class
* Add description of `name` under `Attributes:` of `Extension` class
* Fix docs to link to both attributes using `` :attr:`Extension.name` ``,  `` :attr:`Extension.default_name` ``

File: chainer/training/extension.py [hunk 1]
```diff
@@ -31,10 +31,13 @@ class Extension(object):
             is set to ``(1, 'iteration')`` by default.
         ~Extension.priority: Default priority of the extension. It is set to
             ``PRIORITY_READER`` by default.
+        ~Extension.name: Name of the extension. It is set to
+            ``None`` by default.
```
Comment (kmaehashi) [inline chainer/training/extension.py]: How about adding:

```
This value will be overwritten when registering an extension to a trainer. See :meth:`chainer.training.Trainer.extend` for details.
```

File: chainer/training/trainer.py [hunk 2]
```diff
@@ -192,7 +192,9 @@ def extend(self, extension, name=None, trigger=None, priority=None,
         Args:
             extension: Extension to register.
             name (str): Name of the extension. If it is omitted, the
-                ``default_name`` attribute of the extension is used instead.
+                :attr:`Extension.name` attribute of the extension is used or
+                the :attr:`Extension.default_name` attribute of the extension
+                if ''name'' is also undefined.
```
Comment (kmaehashi) [inline chainer/training/trainer.py]: * `''name''` -> ``` `name` ```
* `also undefined` -> `is set to None or is undefined`

File: chainer/training/extension.py (hunk 1 above)
Comment (Crissman) [reply chainer/training/extension.py]: Done.

File: chainer/training/trainer.py (hunk 2 above)
Comment (Crissman) [reply chainer/training/trainer.py]: Done.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

**Comment by kmaehashi** [comment]
LGTM! Test failures are not related to this PR.

**Pull request merged**